cpp/src/groupby/hash/groupby.cu

```diff
@@ -68,8 +81,248 @@ bool constexpr is_hash_aggregation(aggregation::Kind t) {
+      return std::move(transformed_result->release()[0]);
+    };
+
+    std::for_each(agg_v.begin(), agg_v.end(),
+      [&sparse_results, &dense_results, to_dense_agg_result, transformed_result,
+       &col, i]
+      (auto const& agg) {
+        if (col.type().id() == type_id::STRING and
+            (agg->kind == aggregation::MAX or agg->kind == aggregation::MIN)) {
+          if (agg->kind == aggregation::MAX) {
+            dense_results.add_result(i, agg,
+              transformed_result(aggregation::ARGMAX));
+          }
+          else if (agg->kind == aggregation::MIN) {
+            dense_results.add_result(i, agg,
+              transformed_result(aggregation::ARGMIN));
+          }
+        }
+        else if (sparse_results.has_result(i, agg)) {
+          dense_results.add_result(i, agg, to_dense_agg_result(agg));
+        }
+      });
+  }
+}
+
+/**
+ * @brief Construct hash map that uses row comparator and row hasher on 
+ * @p d_keys table and stores indices
+ */
+template <bool keys_have_nulls>
+auto create_hash_map(table_device_view const& d_keys, bool ignore_null_keys,
+                     cudaStream_t stream = 0)
+{
+  size_type constexpr unused_key{std::numeric_limits<size_type>::max()};
+  size_type constexpr unused_value{std::numeric_limits<size_type>::max()};
+
+  using map_type =
+      concurrent_unordered_map<size_type, size_type,
+                              row_hasher<default_hash, keys_have_nulls>,
+                              row_equality_comparator<keys_have_nulls>>;
+
+  using allocator_type = typename map_type::allocator_type;
+
+  bool const null_keys_are_equal{not ignore_null_keys};
+
+  row_hasher<default_hash, keys_have_nulls> hasher{d_keys};
+  row_equality_comparator<keys_have_nulls> rows_equal{
+      d_keys, d_keys, null_keys_are_equal};
+
+  return map_type::create(compute_hash_table_size(d_keys.num_rows()),
+                            unused_key, unused_value, hasher, rows_equal,
+                            allocator_type(), stream);
+}
+
+/**
+ * @brief Computes all aggregations from @p requests that require a single pass
+ * over the data and stores the results in @p sparse_results
+ */
+template <bool keys_have_nulls, typename Map>
+void compute_single_pass_aggs(table_view const& keys,
+                              std::vector<aggregation_request> const& requests,
+                              experimental::detail::result_cache& sparse_results,
+                              Map& map, bool ignore_null_keys,
+                              cudaStream_t stream)
+{
+  // flatten the aggs to a table that can be operated on by aggregate_row
+  table_view flattened_values;
+  std::vector<aggregation::Kind> aggs;
+  std::vector<size_t> col_ids;
+  std::tie(flattened_values, aggs, col_ids) = flatten_single_pass_aggs(requests);
+
+  // make table that will hold sparse results
+  std::vector<std::unique_ptr<column>> sparse_columns;
+  for (size_t i = 0; i < aggs.size(); i++) {
```

Comment:
`i` isn't used other than to get `aggs[i]`, so I would use a range for.
```suggestion
  for (auto const& agg: aggs) {
```

Reply:
And `auto const& col = flattened_values.column(i);`

Reply:
Basically I need an stl zip iterator. This whole PR, I've tried to look at every loop and think of a way to replace it with an stl algorithm but unable to.

Reply:
> Basically I need an stl zip iterator. This whole PR, I've tried to look at every loop and think of a way to replace it with an stl algorithm but unable to.

Is there a reason why the transform code I posted above would not work here?

Reply:
You can use Thrust zip iterators in STL algorithms.

Reply:
> Is there a reason why the transform code I posted above would not work here?

It worked. I misread it as a unary op transform.